File: 프로그래머스/js/Level 1/폰켓몬.js
function solution(nums) {
  const map = new Map();
  for (let num of nums) {
    if (!map.get(num)) {
      map.set(num, 1);
    } else {
      let newValue = map.get(num) + 1;
      map.set(num, newValue);
    }
  }
  console.log(map.size);
  if (map.size >= nums.length / 2) {
    return nums.length / 2;
  } else {
    return map.size;
  }
}

File: 프로그래머스/python/Level 1/폰켓몬.py
# set 자료구조 
def solution(nums):    
    nums_set = set(nums)
    
    if len(nums_set) > len(nums) / 2:
        return len(nums) / 2
    else:
        return len(nums_set)

# 코드 개선
def solution(nums):
		return min(len(set(nums), len(nums) / 2))

# 딕셔너리 사용
def solution(nums):
    nums_map = {}
    
    for num in nums:
        if not num in nums_map:
            nums_map[num] = True
    
    return min(len(nums_map), len(nums) / 2)